### ISSUE TITLE
Variability in SSD Mixed-Precision Performance

### ISSUE THREAD

**Issue by vuiseng9** [openvinotoolkit/nncf#131]
We observed large variability in performance with SSD300(VGG) when we tried different combination of precision for weight and activation in test mode. From the collected number below, the best and worst are about 15X gap, inference per batch (size: 128) is 30secs for Int2 weights and Int8 activation for the worst case. The performance should impact fine-tuning mode as well.

NNCF Version: Develop branch with commit 2a681b897f992a898dca463b541ee12d9c9899fe
Similar observations with v1.4
Baseline config: https://github.com/openvinotoolkit/nncf_pytorch/blob/develop/examples/object_detection/configs/ssd300_vgg_voc_int8.json
Platform: V100 GPU

|     weights    |     activations    |     detection elapse                                                                                  |
|----------------|--------------------|-------------------------------------------------------------------------------------------------------|
|     8          |     8              |     Detect for batch: 8/39 **1.847s** Detect for batch: 9/39 **1.844s** Detect for batch: 10/39 **1.876s**        |
|     8          |     4              |     Detect for batch: 8/39 **8.792s** Detect for batch: 9/39 **9.021s** Detect for batch: 10/39 **8.928s**        |
|     8          |     2              |     Detect for batch: 8/39 **17.09s** Detect for batch: 9/39 **17.80s** Detect for batch: 10/39 **17.87s**     |
|     4          |     8              |     Detect for batch: 8/39 **2.283s** Detect for batch: 9/39 **2.105s** Detect for batch: 10/39 **2.296s**        |
|     4          |     4              |     Detect for batch: 8/39 **8.285s** Detect for batch: 9/39 **9.583s** Detect for batch: 10/39 **7.425s**        |
|     4          |     2              |     Detect for batch: 8/39 **11.31s** Detect for batch: 9/39 **11.85s** Detect for batch: 10/39 **12.61s**    |
|     2          |     8              |     Detect for batch: 8/39 **29.40s** Detect for batch: 9/39 **30.75s** Detect for batch: 10/39 **29.30s**     |
|     2          |     4              |     Detect for batch: 8/39 **5.684s** Detect for batch: 9/39 **5.703s** Detect for batch: 10/39 **5.539s**        |
|     2          |     2              |     Detect for batch: 8/39 **5.954s** Detect for batch: 9/39 **6.040s** Detect for batch: 10/39 **6.159s**        |

**Comment by vshampor** [comment]
@vuiseng9 I guess the question is still the same - does the slowdown happen in the feature extraction stage or in the detection output stage?

**Comment by vuiseng9** [comment]
@vshampor the bottleneck is in the detection output stage. Pls find the following details from torch profiler. When we change to a different combination of precision, the number of calls of these functions ```is_nonzero, item, _local_scalar_dense, lt``` vary, in the case of W2W8, they are the highest among the experiments, and hence the accumulated elapsed time. From the chrome tracing of the profiler dump, i believe these functions are mainly called in the routine [here](https://github.com/openvinotoolkit/nncf_pytorch/blob/develop/examples/object_detection/layers/functions/detection.py#L77-L144).

p/s: Reduced batch size to batch size to 8 from 128 above and profiled with **W8A8** and **W2A8** mode.

**W8A8**
```
Detect for batch: 5/619 0.734s
------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                                  Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
model_inference                       46.28%           155.953ms        98.89%           333.245ms        333.245ms        41.83%           333.249ms        333.249ms        1                
DetectionOutputFunction               11.45%           38.571ms         35.45%           119.455ms        119.455ms        15.00%           119.457ms        119.457ms        1                
conv2d                                0.11%            370.210us        2.14%            7.226ms          206.460us        5.61%            44.685ms         1.277ms          35               
convolution                           0.08%            275.736us        2.03%            6.856ms          195.883us        5.57%            44.363ms         1.268ms          35               
_convolution                          0.94%            3.174ms          1.95%            6.580ms          188.005us        5.53%            44.077ms         1.259ms          35               
cudnn_convolution                     0.60%            2.022ms          0.60%            2.022ms          57.780us         4.81%            38.289ms         1.094ms          35               
index                                 6.10%            20.546ms         6.10%            20.546ms         50.483us         2.59%            20.669ms         50.783us         407              
NMSFunction                           2.96%            9.961ms          4.97%            16.735ms         104.592us        2.10%            16.752ms         104.697us        160              
item                                  1.28%            4.317ms          9.35%            31.501ms         44.242us         1.78%            14.158ms         19.884us         712              
QuantizeSymmetric                     1.94%            6.524ms          2.87%            9.663ms          134.215us        1.67%            13.285ms         184.520us        72               
------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 336.978ms
CUDA time total: 796.619ms
```
**W2A8**
```
Detect for batch: 5/619 14.744s
------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Name                                  Self CPU total %  Self CPU total   CPU total %      CPU total        CPU time avg     CUDA total %     CUDA total       CUDA time avg    Number of Calls  
------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
model_inference                       4.91%            189.784ms        99.91%           3.860s           3.860s           30.60%           3.860s           3.860s           1                
DetectionOutputFunction               27.00%           1.043s           93.49%           3.612s           3.612s           28.63%           3.612s           3.612s           1                
!! is_nonzero                            7.85%            303.371ms        42.54%           1.643s           26.860us         13.29%           1.676s           27.395us         61184            
!! item                                  12.73%           491.890ms        35.14%           1.358s           22.080us         10.63%           1.341s           21.802us         61488            
!! _local_scalar_dense                   22.41%           865.753ms        22.41%           865.753ms        14.080us         6.55%            825.974ms        13.433us         61488            
!! lt                                    17.74%           685.176ms        17.74%           685.176ms        11.199us         5.99%            755.024ms        12.340us         61184            
!! select                                2.64%            102.085ms        2.64%            102.085ms        3.398us          0.79%            99.940ms         3.327us          30041            
NMSFunction                           0.77%            29.609ms         2.07%            79.925ms         499.529us        0.63%            79.915ms         499.470us        160              
conv2d                                0.01%            401.252us        0.25%            9.499ms          271.386us        0.37%            46.393ms         1.326ms          35               
convolution                           0.01%            301.354us        0.24%            9.097ms          259.922us        0.37%            46.066ms         1.316ms          35               
------------------------------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  ---------------  
Self CPU time total: 3.863s
CUDA time total: 12.615s
```
**Snippet of profiling:**
```python
# examples/object_detection/eval.py

def predict_detections(data_loader, device, net):
    num_batches = len(data_loader)
    all_detections = []
    timer = Timer()
    for batch_ind, (ims, _gts, hs, ws) in enumerate(data_loader):
        x = ims.to(device)
        hs = x.new_tensor(hs).view(-1, 1)
        ws = x.new_tensor(ws).view(-1, 1)

        timer.tic()
        with profiler.profile(record_shapes=True, use_cuda=True) as prof:
            with profiler.record_function("model_inference"):
                batch_detections = net(x)

        top_k = batch_detections.size(2)
        batch_detections = batch_detections.view(-1, top_k, 7)
        detect_time = timer.toc(average=False)

        batch_detections[..., 3] *= ws
        batch_detections[..., 5] *= ws
        batch_detections[..., 4] *= hs
        batch_detections[..., 6] *= hs

        all_detections.append(batch_detections.cpu())
        logger.info('Detect for batch: {:d}/{:d} {:.3f}s'.format(batch_ind + 1, num_batches, detect_time))

        if batch_ind+1 == 5:
            logger.info(prof.key_averages().table(sort_by="cuda_time_total", row_limit=10))
            prof.export_chrome_trace("ssd300_vgg_voc_W2A8_BS8_trace.json")

    if all_detections:
        return torch.cat(all_detections)
    return None  # No predictions
```

**Comment by AlexKoff88** [comment]
@vuiseng9, thanks for investigating this. My guess is that the problem is caused by the discrete nature of confidence values as one of the inputs into the detection output. They get higher than the confidence threshold and cause more computations to match ground truth and detected objects.  To verify this hypothesis we just need to keep confidence Conv layers in FP32. All the rest can be INT2/INT4/INT8.

I see two solutions here:
1. Adjust the confidence threshold. Now it is not clear how to do this automatically.
2. Keep these layers in high precision, at least, in INT8.

**Comment by vuiseng9** [comment]
+ @vanyalzr

**Comment by SKholkin** [comment]
I've benchmarked ssd300 and got results. 
int8: Detect for batch: 35/248 0.399s
Mixed precision:  Detect for batch: 3/248 6.338s
Also for ssd512 difference between time of int8 inference and mixed precision inference is not that big:
int8: Detect for batch: 30/310 0.473s
Mixed precision: Detect for batch: 15/310 1.344s
We continue to investigate
NNCF Version: Develop branch with [ this commit](https://github.com/openvinotoolkit/nncf_pytorch/commit/db8c09d8a1f9efb6c22150ba285ce65bc879d88c) 
Platform: V100 GPU

**Comment by vuiseng9** [comment]
The variability is an expected behavior. When we quantize SSD to lower precision, the confidence prediction is distorted and results in increase of spurious bboxes being fed into the NMS, and hence longer detection elapse. This variability will be minimized after fine-tuning where the network states are preserved to pretrained states. 

Nevertheless, for mixed-precision search where fine-tuning is less preferred in each search iteration due to variability of inference time. The suggestion is to evaluate loss instead of mAP to bypass post-processing of bboxes to achieve constant search time per episode.
